### PR TITLE
Sean - L1233_elec_water

### PR DESCRIPTION
**Pull request by swd-turner**
#310

### CODE REVIEW

File: R/constants.R [hunk 1]
```diff
@@ -95,6 +95,7 @@ CONV_HA_BM2 <- 1e-5
 CONV_THA_KGM2 <- 0.1   # tons C/ha -> kg C/m2
 CONV_GG_TG <- 0.001 # gigagrams to tegagrams
 CONV_TST_TG <- 0.000907 # thousand short tons to Tg
+CONV_MWH_GJ <- 3.6
```
Comment (bpbond) [inline R/constants.R]: Megawatt hours to gigajoules?

File: R/zchunk_L1233.Elec_water.R [hunk 2]
```diff
@@ -1,27 +1,26 @@
...
 #' a vector of output names, or (if \code{command} is "MAKE") all
 #' the generated outputs: \code{L1233.out_EJ_R_elec_F_tech_Yh_cool}, \code{L1233.in_EJ_R_elec_F_tech_Yh_cool}, \code{L1233.wdraw_km3_R_elec}, \code{L1233.wcons_km3_R_elec}, \code{L1233.shrwt_R_elec_cool_Yf}. The corresponding file in the
 #' original data system was \code{L1233.Elec_water.R} (water level1).
-#' @details Describe in detail what this chunk does.
+#' @details Categorizes electiricty generating technologies by water type, and computes water withdrawals and consumption.
 #' @importFrom assertthat assert_that
 #' @importFrom dplyr filter mutate select
 #' @importFrom tidyr gather spread
-#' @author YourInitials CurrentMonthName 2017
-#' @export
```
Comment (bpbond) [inline R/zchunk_L1233.Elec_water.R]: Remove `@export` tag on final commit

File: R/zchunk_L1233.Elec_water.R [hunk 3]
```diff
@@ -39,89 +38,301 @@ module_water_L1233.Elec_water_DISABLED <- function(command, ...) {
...
     enduse_fuel_aggregation <- get_data(all_data, "energy/enduse_fuel_aggregation")
-    L101.en_bal_EJ_ctry_Si_Fi_Yh_full <- get_data(all_data, "L101.en_bal_EJ_ctry_Si_Fi_Yh_full")
-    L1231.in_EJ_R_elec_F_tech_Yh <- get_data(all_data, "L1231.in_EJ_R_elec_F_tech_Yh")
-    L1231.out_EJ_R_elec_F_tech_Yh <- get_data(all_data, "L1231.out_EJ_R_elec_F_tech_Yh")
     A23.CoolingSystemShares_RG3 <- get_data(all_data, "water/A23.CoolingSystemShares_RG3")
     elec_tech_water_map <- get_data(all_data, "water/elec_tech_water_map")
     Macknick_elec_water_m3MWh <- get_data(all_data, "water/Macknick_elec_water_m3MWh")
 
+    L101.en_bal_EJ_ctry_Si_Fi_Yh_full <-
+      get_data(all_data, "temp-data-inject/L101.en_bal_EJ_ctry_Si_Fi_Yh_full") %>%
```
Comment (bpbond) [inline R/zchunk_L1233.Elec_water.R]: I think if you merge `master` this is now available; change and remove from `temp-data-inject` please

File: R/zchunk_L1233.Elec_water.R [hunk 4]
```diff
@@ -39,89 +38,301 @@ module_water_L1233.Elec_water_DISABLED <- function(command, ...) {
...
-      add_title("descriptive title of data") %>%
-      add_units("units") %>%
-      add_comments("comments describing how data generated") %>%
-      add_comments("can be multiple lines") %>%
+    # STEP 1. DOWNSCALE WATER COOLING SYSTEM SHARES FROM 14 REGIONS to 201 COUNTRIES
+
+    # RENAME INTERMEDIATE FUELS AND AGGREGATE BY UPDATED FUEL TYPE
+    L101.en_bal_EJ_ctry_Si_Fi_Yh_full %>%
+      filter(sector == "out_electricity generation") %>%
+      left_join(select(enduse_fuel_aggregation, fuel, electricity), by = "fuel") %>%
```
Comment (bpbond) [inline R/zchunk_L1233.Elec_water.R]: More restrictive join possible?

File: R/zchunk_L1233.Elec_water.R [hunk 5]
```diff
@@ -39,89 +38,301 @@ module_water_L1233.Elec_water_DISABLED <- function(command, ...) {
...
+      select(-fuel) %>%
+      rename(fuel = electricity) -> L1233.out_EJ_ctry_elec_Fi_Yh_full
+
+    L1233.out_EJ_ctry_elec_Fi_Yh_full %>%
+      filter(year %in% HISTORICAL_YEARS) %>%
+      group_by(iso, year, fuel) %>%
+      summarise(value = sum(value)) %>%
+      ungroup() %>%
+      filter(is.na(fuel) == F) %>%
```
Comment (bpbond) [inline R/zchunk_L1233.Elec_water.R]: `FALSE`

File: R/zchunk_L1233.Elec_water.R [hunk 6]
```diff
@@ -39,89 +38,301 @@ module_water_L1233.Elec_water_DISABLED <- function(command, ...) {
...
+      mutate(sector = "electricity generation") -> L1233.out_EJ_ctry_elec_F_Yh
```
Comment (bpbond) [inline R/zchunk_L1233.Elec_water.R]: A few more comments would be useful. Why is this happening?

File: R/zchunk_L1233.Elec_water.R [hunk 7]
```diff
@@ -39,89 +38,301 @@ module_water_L1233.Elec_water_DISABLED <- function(command, ...) {
...
+
+    # INTERPOLATE A23.CoolingSystemShares_RG3 FOR HISTORICAL YEARS AND HISTORICAL + FUTURE YEARS
+    A23.CoolingSystemShares_RG3 %>%
+      gather(year, value, -region_GCAM3, -plant_type, -cooling_system, -water_type) %>%
+      mutate(year = as.integer(year)) -> A23.CoolingSystemShares_RG3_LF
+
+    INTERPOLATED_YEARS <- HISTORICAL_YEARS[HISTORICAL_YEARS %in%
```
Comment (bpbond) [inline R/zchunk_L1233.Elec_water.R]: In general save ALL CAPS for constants

File: R/zchunk_L1233.Elec_water.R [hunk 8]
```diff
@@ -39,89 +38,301 @@ module_water_L1233.Elec_water_DISABLED <- function(command, ...) {
...
+                                              unique(A23.CoolingSystemShares_RG3_LF$year) == F]
+    EXTRAPOLATED_YEARS <- c(HISTORICAL_YEARS, FUTURE_YEARS)[c(HISTORICAL_YEARS, FUTURE_YEARS) %in%
+                                                              unique(A23.CoolingSystemShares_RG3_LF$year) == F]
+
+    setNames(data.frame(matrix(ncol = length(INTERPOLATED_YEARS),
+                               nrow = nrow(A23.CoolingSystemShares_RG3))),
+             as.character(INTERPOLATED_YEARS)) %>%
```
Comment (bpbond) [inline R/zchunk_L1233.Elec_water.R]: What's happening here? Why go matrix -> data frame -> tibble?

File: R/zchunk_L1233.Elec_water.R [hunk 9]
```diff
@@ -39,89 +38,301 @@ module_water_L1233.Elec_water_DISABLED <- function(command, ...) {
...
+      mutate(value = approx_fun(year, value, rule = 1)) %>%
+      ungroup() -> A23.CoolingSystemShares_RG3_LF_interp
+
+    setNames(data.frame(matrix(ncol = length(EXTRAPOLATED_YEARS),
+                               nrow = nrow(A23.CoolingSystemShares_RG3))),
+             as.character(EXTRAPOLATED_YEARS)) %>%
+      as_tibble() %>%
+      bind_cols(A23.CoolingSystemShares_RG3) %>%
+      gather(year, value, -region_GCAM3, -plant_type, -cooling_system, -water_type) %>%
+      mutate(year = as.integer(year)) %>%
```
Comment (bpbond) [inline R/zchunk_L1233.Elec_water.R]: You keep doing this. Why not convert A23.CoolingSystemShares_RG3's `year` column to integer once, when it's read in?

File: R/zchunk_L1233.Elec_water.R [hunk 10]
```diff
@@ -39,89 +38,301 @@ module_water_L1233.Elec_water_DISABLED <- function(command, ...) {
...
+      arrange(year) %>%
+      group_by(region_GCAM3, plant_type, cooling_system, water_type) %>%
+      mutate(value = approx_fun(year, value, rule = 1)) %>%
+      ungroup() -> A23.CoolingSystemShares_RG3_LF_future
+
+    # WRITE OUT REGIONAL LEVEL COOLING INFORMATION TO ALL COUNTRIES...
+    # ... AND MATCH GCAM region_GCAM3 INTO TIBBLE
+    elec_tech_water_map %>%
+      semi_join(calibrated_techs, by = c("sector", "fuel", "technology")) %>%
```
Comment (bpbond) [inline R/zchunk_L1233.Elec_water.R]: Some more comments would be good. Why the `semi_join` here?

File: R/zchunk_L1233.Elec_water.R [hunk 11]
```diff
@@ -39,89 +38,301 @@ module_water_L1233.Elec_water_DISABLED <- function(command, ...) {
...
+    elec_tech_water_map_cal %>%
+      repeat_add_columns(select(L1233.out_EJ_ctry_elec_F_Yh, iso) %>% unique()) %>%
+      left_join_error_no_match(select(iso_GCAM_regID, iso, region_GCAM3), by = "iso") %>%
+      left_join(A23.CoolingSystemShares_RG3_LF_interp,
+                by = c("cooling_system", "water_type", "plant_type", "region_GCAM3")) -> L1233.weights_ctry_elec_F_Yh_cool
+
+    # ADD TECHNOLOGIES WITH NO COOLING FOR ALL HISTORICAL YEARS
+    L1233.weights_ctry_elec_F_Yh_cool %>%
+      filter(plant_type == "no cooling") %>%
+      mutate(value = 1) %>%
```
Comment (bpbond) [inline R/zchunk_L1233.Elec_water.R]: ?

File: R/zchunk_L1233.Elec_water.R [hunk 12]
```diff
@@ -39,89 +38,301 @@ module_water_L1233.Elec_water_DISABLED <- function(command, ...) {
...
+
+
+    # STEP 2. MULTIPLY SHARES BY COUNTRY LEVEL GENERATION TO GET ELECTRICITY...
+    # ... OUTPUT BY GENERATING AND COOLING TECHNOLOGIES
+    L1233.weights_ctry_elec_F_Yh_cool_withNoCool %>%
+      left_join(L1233.out_EJ_ctry_elec_F_Yh, by = c("iso", "year", "fuel", "sector")) %>%
+      rename(gen = value) %>%
+      mutate(output = share * gen) %>%
+      select(-share, -gen) %>%
+      mutate(output = if_else(is.na(output) == T, 0, output)) -> L1233.output_ctry_elec_F_Yh_cool
```
Comment (bpbond) [inline R/zchunk_L1233.Elec_water.R]: `is.na(output), 0, output`

File: R/zchunk_L1233.Elec_water.R [hunk 13]
```diff
@@ -39,89 +38,301 @@ module_water_L1233.Elec_water_DISABLED <- function(command, ...) {
...
+      repeat_add_columns(unique(select(iso_GCAM_regID, GCAM_region_ID))) %>%
+      select(GCAM_region_ID, sector, fuel, technology, cooling_system, water_type, plant_type) %>%
+      left_join(L1233.weights_R_elec_F_Yh_cool,
+                by = c("GCAM_region_ID", "sector", "fuel", "technology", "cooling_system", "water_type")) %>%
+      rename(output_cool = output) %>%
+      left_join(L1233.weights_R_elec_F_Yh,
+                by = c("GCAM_region_ID", "sector", "fuel", "technology", "year")) %>%
+      mutate(share = output_cool / output) %>% select(-output_cool, -output) %>%
+      arrange(GCAM_region_ID, sector, fuel, technology, cooling_system, water_type, plant_type, year) %>%
+      mutate(share = if_else(year == tail(HISTORICAL_YEARS, 1) & is.na(share), -1, share)) ->
```
Comment (bpbond) [inline R/zchunk_L1233.Elec_water.R]: `last(HISTORICAL_YEARS)`

File: R/zchunk_L1233.Elec_water.R [hunk 14]
```diff
@@ -39,89 +38,301 @@ module_water_L1233.Elec_water_DISABLED <- function(command, ...) {
...
+      L1233.shares_R_elec_F_tech_Yh_cool
+
+    ## REMOVE NA VALUES BY PULLING FROM YEAR AHEAD IN EACH GROUP
+    for(i in (nrow(L1233.shares_R_elec_F_tech_Yh_cool) - 1) : 1){
```
Comment (bpbond) [inline R/zchunk_L1233.Elec_water.R]: Ugh! See `lag` and `lead` and use if possible.

File: R/zchunk_L1233.Elec_water.R [hunk 15]
```diff
@@ -39,89 +38,301 @@ module_water_L1233.Elec_water_DISABLED <- function(command, ...) {
...
+      summarise(value = sum(value)) %>%
+      ungroup() %>%
+      filter(water_type != "none") -> L1233.wcons_km3_R_elec
+
+    # STEP 8: COMPUTE REGIONAL-LEVEL FUTURE SHARE WEIGHTS...
+    # ... BASED ON GCAM 3.0, 14-REGION ASSUMPTIONS. NOTE: USES ...
+    # ... REPRESENTATIVE COUNTRIES RATHER THAN WEIGHTED AVERAGES ...
+    # ... (COUNTRY WITH MOST ELEC AS THE REPRESENTATIVE)
+    L1233.out_EJ_ctry_elec_Fi_Yh_full %>%
+      filter(year == tail(HISTORICAL_YEARS, 1)) %>%
```
Comment (bpbond) [inline R/zchunk_L1233.Elec_water.R]: `dplyr::last`

File: R/zchunk_L1233.Elec_water.R [hunk 16]
```diff
@@ -39,89 +38,301 @@ module_water_L1233.Elec_water_DISABLED <- function(command, ...) {
...
+
+
+    # STEP 9: FILL OUT TABLE OF ALL TECHNOLOGIES TO MATCH IN SHARES
+    elec_tech_water_map %>%
+      select(sector, fuel, technology, cooling_system, water_type, plant_type) %>%
+      repeat_add_columns(select(iso_GCAM_regID, GCAM_region_ID) %>% unique()) %>%
+      arrange(GCAM_region_ID) %>%
+      left_join(select(L1233.R_iso_RG3, -value, -iso),
+                by = "GCAM_region_ID") %>%
+      mutate(plant_type = sub( "\\ \\(CCS\\)", "", plant_type)) %>%
```
Comment (bpbond) [inline R/zchunk_L1233.Elec_water.R]: ?

File: R/zchunk_L1233.Elec_water.R [hunk 17]
```diff
@@ -39,89 +38,301 @@ module_water_L1233.Elec_water_DISABLED <- function(command, ...) {
...
-      # typical flags, but there are others--see `constants.R`
+      add_precursors("common/iso_GCAM_regID",
+                     "energy/calibrated_techs",
+                     "energy/enduse_fuel_aggregation",
+                     "temp-data-inject/L101.en_bal_EJ_ctry_Si_Fi_Yh_full",
+                     "temp-data-inject/L1231.in_EJ_R_elec_F_tech_Yh",
+                     "temp-data-inject/L1231.out_EJ_R_elec_F_tech_Yh",
+                     "water/A23.CoolingSystemShares_RG3",
+                     "water/elec_tech_water_map",
+                     "water/Macknick_elec_water_m3MWh") %>%
```
Comment (bpbond) [inline R/zchunk_L1233.Elec_water.R]: Have you checked that _all_ outputs use _all_ inputs, as declared here?

File: inst/extdata/common/iso_GCAM_regID.csv [hunk 18]
```diff
@@ -1,244 +1,244 @@
-# File:iso_GCAM_regID.csv,,,
```
Comment (bpbond) [inline inst/extdata/common/iso_GCAM_regID.csv]: Don't replace existing file unless you have a good reason please.

File: R/zchunk_L1233.Elec_water.R (hunk 2 above)
Comment (swd-turner) [inline R/zchunk_L1233.Elec_water.R]: Weird... my copy has the export tag removed (and initials filled in). I guess I screwed something up when I was resolving conflicts with master!

Comment (bpbond) [inline R/zchunk_L1233.Elec_water.R]: @abigailsnyder huh, had the same thing happen today. Not sure what's going on.

File: R/zchunk_L1233.Elec_water.R (hunk 6 above)
Comment (swd-turner) [inline R/zchunk_L1233.Elec_water.R]: Ok - I must confess I lost the thread of what was happening in places due to the sheer size of the script. Will comb through and add in more comments!

File: R/zchunk_L1233.Elec_water.R (hunk 8 above)
Comment (swd-turner) [inline R/zchunk_L1233.Elec_water.R]: Probably not necessary actually. I'm setting up a blank dataframe with headers equal to years I want to insert into a tibble. I then gather it and bind to the existing tibble before using the approx_fun to interpolate them. Is there a smarter way of inserting blank years (i.e., with NA values and repeating for all groups) into a tibble before interpolating? (As I write this out I'm thinking repeat_add_columns might do the job...).

File: R/zchunk_L1233.Elec_water.R (hunk 14 above)
Comment (swd-turner) [inline R/zchunk_L1233.Elec_water.R]: Agree, this is horrible. `lag `and `lead` won't cut it as there's dependency. It's more of a fill operation. I'll spend a bit more time on it to see if I can get rid of the for loop!

File: R/zchunk_L1233.Elec_water.R (hunk 8 above)
Comment (bpbond) [inline R/zchunk_L1233.Elec_water.R]: I think you want `tidyr::complete`? It kicks ass.

File: R/zchunk_L1233.Elec_water.R (hunk 14 above)
Comment (bpbond) [inline R/zchunk_L1233.Elec_water.R]: I think this is a one-liner with `tidyr::fill`.

File: R/zchunk_L1233.Elec_water.R (hunk 2 above)
Comment (abigailsnyder) [inline R/zchunk_L1233.Elec_water.R]: I think I figured it out - I think the red text with the minus before the poundsign indicates that line was removed. Looking at the full diff on this script and on mine yesterday, right below the red is a green line with a plus out front and the correct author information. Perhaps this indicates that the blank author and export lines were removed, and the correct author info line was added?

Comment (abigailsnyder) [inline R/zchunk_L1233.Elec_water.R]: actually, I misspoke. The + correct information is a little bit more below - the github online diff view clumps the remove author skeleton, remove export, and the remove the module_DISABLED = function... lines. Then below _that_ is the green addition line with the correct author information. 

OR everything is broken and nothing makes sense and this is just a new and exciting problem for us to look at every time.

Comment (abigailsnyder) [inline R/zchunk_L1233.Elec_water.R]: fwiw, I decided to be a little nebby and checked that the same sequence of three removed lines followed by the addition of a line with correct author information occurs in one of the other open pull requests at the moment. It did.

Comment (bpbond) [inline R/zchunk_L1233.Elec_water.R]: Yeah, so I was brain-dead with this comment then. Sorry.

File: R/zchunk_L1233.Elec_water.R (hunk 3 above)
Comment (swd-turner) [inline R/zchunk_L1233.Elec_water.R]: Ok - this one isn't available yet, but the other two temp-inject files are :+1:

**Comment by bpbond** [comment]
@swd-turner sorry I just meant to fix NAMESPACE conflict but appear to have broken your PR. Let me know and I can fix, or not.

File: R/zchunk_L1233.Elec_water.R (hunk 4 above)
Comment (swd-turner) [inline R/zchunk_L1233.Elec_water.R]: Actually this should technically be a restricted join because there are matching entries for all fuels. The problem is some of the entries in the file being joined are already labelled NA (i.e., before the join). The `left_join_error_no_match` function sees the NAs in the output and assumes incorrectly that these are failed matches... I'm wondering if `left_join_error_no_match` should be relaxed to allow for this type of case.

Comment (bpbond) [inline R/zchunk_L1233.Elec_water.R]: I think the solution would not be to relax the existing function, but allow for different behavior, something we've talked about but not got around to doing. E.g. left_join and assert that number of rows shouldn't change; left_join and assert no NAs; left_join and assert both (the behavior of the current `left_join_error_no_match`).

Comment (swd-turner) [inline R/zchunk_L1233.Elec_water.R]: ...the other option of course is to simply relabel the data in the file being joined as "undefined", or similar, rather than <NA>.

**Comment by swd-turner** [comment]
@bpbond no worries I'll try to fix it up when I commit and merge my changes later.

File: R/zchunk_L1233.Elec_water.R (hunk 8 above)
Comment (swd-turner) [inline R/zchunk_L1233.Elec_water.R]: Works a charm. My life is now `tidyr::complete`.

File: inst/extdata/common/iso_GCAM_regID.csv (hunk 18 above)
Comment (swd-turner) [inline inst/extdata/common/iso_GCAM_regID.csv]: Not entirely sure what the deal is with this. I  don't think I replaced this file... maybe opened it in excel and induced new commas or something.

File: R/zchunk_L1233.Elec_water.R (hunk 4 above)
Comment (rplzzz) [inline R/zchunk_L1233.Elec_water.R]: Do _not_ do that.  It creates all manner of headaches for users downstream who now have to figure out how to identify missing values in your data set.  On top of that, it forces all the data in the column to type character, creating more headaches.

Comment (swd-turner) [inline R/zchunk_L1233.Elec_water.R]: Good points--I didn't think that through. I just stuck with `left_join` here so we're good.

File: inst/extdata/common/iso_GCAM_regID.csv (hunk 18 above)
Comment (rplzzz) [inline inst/extdata/common/iso_GCAM_regID.csv]: Looks like it got Windows line endings, which are mostly harmless, but we should revert this change anyhow.

File: inst/extdata/water/elec_tech_water_map.csv [hunk 19]
```diff
@@ -0,0 +1 @@
...
electricity,wind,wind_storage,electricity,wind,wind_storage,no cooling,none,none,electricity generation,wind,wind_storage,onshore wind resource
electricity,solar,PV_storage,electricity,solar,PV_storage,no cooling,none,fresh,electricity generation,solar PV,PV_storage,global solar resource
electricity,solar,CSP_storage,elec_CSP_storage,CSP_storage,CSP_storage (recirculating),CSP,recirculating,fresh,electricity generation,solar CSP,CSP_storage,global solar resource
electricity,solar,CSP_storage,elec_CSP_storage,CSP_storage,CSP_storage (dry_hybrid),CSP,dry_hybrid,fresh,electricity generation,solar CSP,CSP_storage,global solar resource
electricity,geothermal,geothermal,elec_geothermal,geothermal,geothermal (recirculating),geothermal,recirculating,fresh,electricity generation,geothermal,geothermal,geothermal
electricity,geothermal,geothermal,elec_geothermal,geothermal,geothermal (dry_hybrid),geothermal,dry_hybrid,fresh,electricity generation,geothermal,geothermal,geothermal
elect_td_bld,rooftop_pv,rooftop_pv,elect_td_bld,rooftop_pv,rooftop_pv,no cooling,none,none,,,,distributed_solar
```
Comment (rplzzz) [inline inst/extdata/water/elec_tech_water_map.csv]: This one, on the other hand, has bad line endings that will need to be fixed.

File: R/zchunk_L1233.Elec_water.R (hunk 3 above)
Comment (bpbond) [inline R/zchunk_L1233.Elec_water.R]: Also make sure to remove those two files from `inst/extdata/temp-data-inject/ pls

File: inst/extdata/common/iso_GCAM_regID.csv (hunk 18 above)
Comment (swd-turner) [inline inst/extdata/common/iso_GCAM_regID.csv]: Will do.

File: R/zchunk_L1233.Elec_water.R [hunk 20]
```diff
@@ -39,89 +38,250 @@ module_water_L1233.Elec_water_DISABLED <- function(command, ...) {
...
+      rename(output_cool = output) %>%
+      left_join(L1233.weights_R_elec_F_Yh,
+                by = c("GCAM_region_ID", "sector", "fuel", "technology", "year")) %>%
+      mutate(share = if_else(output + output_cool == 0, NA_real_, output_cool / output)) %>%
+      # ^^ above used to catch possible NaN and set as NA instead so that tidyr::fill works below
+      select(-output_cool, -output) %>%
+      arrange(GCAM_region_ID, sector, fuel, technology, cooling_system, water_type, plant_type, year) %>%
+      group_by(GCAM_region_ID, sector, fuel, technology, cooling_system, water_type, plant_type) %>%
+      fill(share, .direction = "up") %>% ungroup() %>%
+      mutate(share = if_else(is.na(share), -1, share)) ->
```
Comment (bpbond) [inline R/zchunk_L1233.Elec_water.R]: use `replace_na` please

File: R/zchunk_L1233.Elec_water.R [hunk 21]
```diff
@@ -39,89 +38,250 @@ module_water_L1233.Elec_water_DISABLED <- function(command, ...) {
...
+
+    ## STEP 2. MULTIPLY SHARES BY COUNTRY LEVEL GENERATION TO GET ELECTRICITY...
+    ## ... OUTPUT BY GENERATING AND COOLING TECHNOLOGIES
+    L1233.weights_ctry_elec_F_Yh_cool %>%
+      left_join(L1233.out_EJ_ctry_elec_F_Yh, by = c("iso", "year", "fuel", "sector")) %>%
+      # ^^ non-restrictive join required for additional combinations of grouping variables in initial tibble
+      rename(gen = value) %>%
+      mutate(output = share * gen) %>%
+      select(-share, -gen) %>%
+      mutate(output = if_else(is.na(output), 0, output)) -> L1233.output_ctry_elec_F_Yh_cool
```
Comment (bpbond) [inline R/zchunk_L1233.Elec_water.R]: Use `replace_na` please

File: inst/extdata/water/elec_tech_water_map.csv (hunk 19 above)
Comment (swd-turner) [inline inst/extdata/water/elec_tech_water_map.csv]: @rplzzz thanks for reviewing. Is the correct approach here to open in a text editor and change the EOL conversion to Unix?

Comment (rplzzz) [inline inst/extdata/water/elec_tech_water_map.csv]: If your text editor will do that, then yes.  Otherwise, I can post a C program that you can compile and run to take care of it.

**Comment by bpbond** [comment]
Holding on review/approval by @mihejazi

**Comment by swd-turner** [comment]
@bpbond feel free to assign another chunk to me in the meantime.

File: R/zchunk_L1233.Elec_water.R [hunk 22]
```diff
@@ -1,25 +1,24 @@
...
+#' Water consumption and withdrawals for electricity.
 #'
 #' @param command API command to execute
 #' @param ... other optional parameters, depending on command
 #' @return Depends on \code{command}: either a vector of required inputs,
 #' a vector of output names, or (if \code{command} is "MAKE") all
 #' the generated outputs: \code{L1233.out_EJ_R_elec_F_tech_Yh_cool}, \code{L1233.in_EJ_R_elec_F_tech_Yh_cool}, \code{L1233.wdraw_km3_R_elec}, \code{L1233.wcons_km3_R_elec}, \code{L1233.shrwt_R_elec_cool_Yf}. The corresponding file in the
 #' original data system was \code{L1233.Elec_water.R} (water level1).
-#' @details Describe in detail what this chunk does.
+#' @details Categorizes electiricty generating technologies by water type, and computes water withdrawals and consumption.
```
Comment (skim301) [inline R/zchunk_L1233.Elec_water.R]: typo: electricity, by "cooling water type" would be more descriptive.

**Comment by bpbond** [comment]
Huge thanks @skim301 ! @swd-turner I'll take care of these.
